### PR TITLE
implement same_value casting for numpy <-> quadtype

### PR DESCRIPTION
**Pull request by mattip**
Closes numpy/numpy-quaddtype#26 to implement `same_value` casting.

I did not look at quad <-> quad casting. CC @seberg for visibility, as far as I can tell the `NPY_SAME_VALUE_CASTING_FLAG` was sufficient to percolate correctly through the can_cast mechanism.

### CODE REVIEW

File: quaddtype/numpy_quaddtype/src/casts.cpp [hunk 1]
```diff
@@ -401,7 +401,7 @@ to_quad<long double>(long double x, QuadBackendType backend)
 }
 
 template <typename T>
-static NPY_CASTING
+static int
```
Comment (mattip) [inline quaddtype/numpy_quaddtype/src/casts.cpp:404]: Hmm... I think this is OK? I didn't see any new warnings, but there are thousands of build warnings when using sleef and I might have missed it.

File: quaddtype/numpy_quaddtype/src/casts.cpp [hunk 2]
```diff
@@ -771,7 +844,11 @@ add_cast_from(PyArray_DTypeMeta *to)
             .name = "cast_QuadPrec_to_NumPy",
             .nin = 1,
             .nout = 1,
+#if NPY_FEATURE_VERSION > NPY_2_3_API_VERSION
+            .casting = NPY_SAME_VALUE_CASTING,
+#else
             .casting = NPY_UNSAFE_CASTING,
+#endif
```
Comment (mattip) [inline quaddtype/numpy_quaddtype/src/casts.cpp:851]: I am not clear how this is used, doesn't the resolver actually state which casting is supported?

**Comment by ngoldbaum** [comment]
Hmm, that's a good point that we need to wait for numpy releases go add features here when often we want to work on both in tandem.

I'd be for adding a new CI job that depends on numpy `main`. We also want to test against the last stable release too, I think.

I think ideally also you should be able to write this with conditional compilation so it works on older numpy versions, but maybe that's hard in practice. I don't think it's a *huge* deal to raise the minimum requires numpy version here but also I think it's worth looking to see if we can avoid that.

File: quaddtype/numpy_quaddtype/src/casts.cpp [hunk 3]
```diff
@@ -666,6 +670,28 @@ from_quad<long double>(quad_value x, QuadBackendType backend)
     }
 }
 
+template <typename T>
+static inline int
+from_quad_checked(quad_value x, QuadBackendType backend, typename NpyType<T>::TYPE *ret) {
+    *ret = from_quad<typename NpyType<T>::TYPE>(x, backend);
+    quad_value check = to_quad<typename NpyType<T>::TYPE>(*ret, backend);
+    if (backend == BACKEND_SLEEF) {
+        if (check.sleef_value == x.sleef_value) {
```
Comment (juntyr) [inline quaddtype/numpy_quaddtype/src/casts.cpp:679]: Can we just compare sleef values here? Should this use equality with NaN = NaN?

File: quaddtype/tests/test_quaddtype.py [hunk 4]
```diff
@@ -76,6 +76,16 @@ def test_unsupported_astype(dtype):
           np.array(QuadPrecision(1)).astype(dtype, casting="unsafe")
 
 
+def test_same_value_cast():
+    # This will fail if compiled with NPY_TARGET_VERSION NPY<2_4_API_VERSION
```
[the remark this answers is not in the document]
Comment (mattip) [reply quaddtype/tests/test_quaddtype.py:80]: It needs both installed `numpy.__version__ >=2.23` and that numpy_quadtype is built with the 2.24+ C-API

File: quaddtype/numpy_quaddtype/src/casts.cpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 #define PY_ARRAY_UNIQUE_SYMBOL QuadPrecType_ARRAY_API
 #define PY_UFUNC_UNIQUE_SYMBOL QuadPrecType_UFUNC_API
-#define NPY_NO_DEPRECATED_API NPY_2_0_API_VERSION
-#define NPY_TARGET_VERSION NPY_2_0_API_VERSION
+#define NPY_NO_DEPRECATED_API NPY_2_4_API_VERSION
+#define NPY_TARGET_VERSION NPY_2_4_API_VERSION
```
[the remark this answers is not in the document]
Comment (mattip) [reply quaddtype/numpy_quaddtype/src/casts.cpp:4]: > make sure that the target version is defined in (and matches up) with the file that does the actual NumPy import/loading

Good point, that is `quaddtype_main.c`.

> but you cannot use the flag at runtime unless you also check PyArray_RUNTIME_VERSION at runtime

Did you mean `PyArray_GetNDArrayCFeatureVersion() >= NPY_2_4_API_VERSION`?

[the remark this answers is not in the document]
Comment (seberg) [reply quaddtype/numpy_quaddtype/src/casts.cpp:4]: I mean `PyArray_RUNTIME_VERSION >= NPY_2_4_API_VERSION`.

[the remark this answers is not in the document]
Comment (mattip) [reply quaddtype/numpy_quaddtype/src/casts.cpp:4]: How does that work at runtime? It is not exported as a constant from `_multiarray_umath.so`

[the remark this answers is not in the document]
Comment (seberg) [reply quaddtype/numpy_quaddtype/src/casts.cpp:4]: It's injected as a constant into your library, next to the API table.  Dunno if that was the right choice, but there we are :).

[the remark this answers is not in the document]
Comment (seberg) [reply quaddtype/numpy_quaddtype/src/casts.cpp]: `PyArray_RUNTIME_VERSION` is _runtime_!  Not a pre-processor macro.

**Comment by mattip** [comment]
Boosting the comment about taking NAN into account when comparing float values which is still TBD. Also testing is not that extensive for edge cases.
> In the SLEEF branch, I was thinking of `Sleef_icmpeqq1(a, b) | (Sleef_iunordq1(a, a) & Sleef_iunordq1(b, b))`

**Comment by SwayamInSync** [comment]
https://numpy.org/devdocs/reference/generated/numpy.can_cast.html

why this feature isn't get added in `can_cast`?

**Comment by juntyr** [comment]
I can only guess but probably since the other cast modes are type based and same value is value based, such that it can allow safe casting for some values even if their types cannot be safely cast in general?

**Comment by mattip** [comment]
Maybe an oversight? Does it make sense to return False if the dtype did not register astype loops?

**Comment by SwayamInSync** [comment]
@seberg one more query, NumPy's byte-swap step not support `same_value` casting?
